Hudi Test Suite

yanghua opened the pull request:
## What is the purpose of the pull request

*This pull request builds the infrastructure of test suite*

## Brief change log

    - Flexible schema payload generation
    - Different types of workload generation such as inserts, upserts etc
    - Post process actions to perform validations
    - Interoperability of test suite to use HoodieWriteClient and HoodieDeltaStreamer so both code paths can be tested
    - Custom workload sequence generator
    - Ability to perform parallel operations, such as upsert and compaction

## Verify this pull request

This pull request is already covered by existing tests*.


## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

yanghua commented:
Hi @n3nash and @vinothchandar I rebased the original PR #623 (hudi_test_suite_refactor feature branch). However, It seems I can not combine with the old commits. So I created a new draft PR.

lamber-ken commented:
👍

yanghua commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/converter/Converter.java:
```diff
@@ -0,0 +1,33 @@
+ * @param <I> Input Data Type
+ * @param <O> Output Data Type
+ */
+public interface Converter<I, O> extends Serializable {
```
I am still confused like the old PR #991 . We have removed `Converter` before. Shall we introduce it again? @n3nash @vinothchandar

yanghua replied:
@vinothchandar Still want to get an answer of this question.

n3nash replied:
@yanghua This is used for test-suite use-cases.

yanghua replied:
When reviewing #991 I have surveyed the git history, it seems Hudi has `Converter ` before, however, we removed it later. I do not know if it's worth that we reintroduce it just for the test suite.

yanghua commented:
I have pushed a new branch named `hudi_test_suite` and will keep this PR syncing with the branch.
This PR used to discuss. If there is no objection, will delete the old branch `hudi_test_suite_refactor`. WDYT? @vinothchandar @n3nash

n3nash commented:
@yanghua Can we continue to use the same branch `hudi_test_suite_refactor` please ? You can continue to PR against that branch. 
Secondly, I think we should name the package - `hudi-test-suite` instead of `hudi-end-to-end-tests`, it's more concise, WDYT ?

Also, can we execute on the following plan :

1) Please go over all the changes in the existing PR - I've addressed most of the comments. 
2) Let's rename the package - like you've already done
3) Both of us can start to use the test suite and run them locally for 2-3 days. This will give us confidence whether the test suite is ready to test large features.
4) Let's merge this first version of the test suite to master so folks can start using it
5) Both of us can continue to enhance the test suite - a) You can add some of the things that you mentioned earlier like DIstributedTestDataSource etc b) I want to add more use-cases

Let me know if this sounds fair to you.

yanghua commented:
> @yanghua Can we continue to use the same branch `hudi_test_suite_refactor` please ? You can continue to PR against that branch. I've also fixed the issue with POM that was causing the test case failure, you may want to rebase this PR against that branch.

I am comfortable that no matter use which branch. The reason why I pushed a new `hudi_test_suite` branch is [here](https://github.com/apache/incubator-hudi/pull/991#issuecomment-559362785). I pinged you and gave a time gap to wait for the suggestion. And I got the suggestion from @vinothchandar . If you can fix the conflicts based on master brach, I can return back to the `hudi_test_suite_refactor` branch. Another option, you can cherry pick your commits about pom issues into my new branch and I can rename the branch name.

> Secondly, I think we should name the package - `hudi-test-suite` instead of `hudi-end-to-end-tests`, it's more concise, WDYT ?

Ok, agree. After coming to an agreement on the branch problem, we can fix the module naming problem.

> 
> Also, can we execute on the following plan :
> 
> 1. Please go over all the changes in the existing PR - I've addressed most of the comments.

There were too many comments in the old PR #991. Reviewing that PR will spend too much time, let's review the new PR to eliminate interruptions of outdate comments. It's a more effective way. WDYT?

> 2. Let's rename the package - like you've already done

OK, agree.

> 3. Both of us can start to use the test suite and run them locally for 2-3 days. This will give us confidence whether the test suite is ready to test large features.

OK, will try it.

> 4. Let's merge this first version of the test suite to master so folks can start using it

I will let @vinothchandar make the decision to see if it's the right time to merge it into the master branch.

> 5. Both of us can continue to enhance the test suite - a) You can add some of the things that you mentioned earlier like DIstributedTestDataSource etc b) I want to add more use-cases
> 

OK

> Let me know if this sounds fair to you.

First of all, you can see I kept your code and your git commits. Something I have done is just squashed them into one to fix the conflicts.

Some things may be caused by our insufficiency in communication. If we can respond to each other's information more efficiently, then maybe our work will be more efficient.

Personally, I try to respond to my Github messages every day, except on weekends or long holidays. @n3nash  Do you pay more attention to Slack? If so, I can communicate with you via Slack. Or can we create a test suite Slack group? However, IMO, communicate via Github is a better way because we can discuss code.

vinothchandar commented:
@n3nash is there an ETA for fixing conflicts and having a buildable branch here? I would also like to understand if you are actively working on this.. This is currently gating the ability to make a lot of large changes we want to do for this next release. Can we discuss this over in the next weekly meeting?

And I second @yanghua 's opinion that , for this specific issue, it would be good to have a faster turnaround time and keep conversations on GH. I am still unclear on ownership here.. I think its safe to say vinoyang would be blocked until @n3nash can resolve all comments and conflicts and get a buildable feature branch? or is that done already? 

On merging, I personally am not comfortable merging until we can run complex tests for atleast a week and demonstrate its usefulness.

n3nash commented:
@yanghua The branch was fixed a couple of days ago and I pushed the changes to https://github.com/apache/incubator-hudi/tree/hudi_test_suite_refactor. Are you referring to the branch being buildable as conflicts ? Just want to make sure that's the only thing required. 

I am fine with your suggestion on reviewing the new PR, agree that will help eliminate dangling comments from before. 

@vinothchandar Last week was thanksgiving and many people are on vacation, including me :)
As far as buildable branch stands, I pushed the changes to make them buildable few days ago. I mentioned this exact same thing in my last comment. It was slightly complex due to weird Hive POM issues and took a longer turnaround time. 

@vinothchandar I'm not sure what is your confusion on the ownership here, which you have repeatedly mentioned. We've already discussed this in the weekly meeting and I've addressed most of the comments and some open questions still remained, pending item was making the branch buildable which is done now. 
As @yanghua is suggesting, we might want to take another pass at the PR to resolve any of those open questions (or take a look at the old PR if one prefers). Once that is done, we want to run these tests for some time - we could do it for a week and @yanghua and I can build further on top of these changes. 
To build confidence, I've also suggested @yanghua to start trying to use the framework. This will help understand and demonstrate it's usefulness (along with running these tests for sometime) before we decide to merge.

vinothchandar commented:
Great to see this making progress.. 

@n3nash Not to nit, but just coz you raised these points, I feel obligated to layout some facts


>Last week was thanksgiving and many people are on vacation, including me :)

This was the last [update](https://github.com/apache/incubator-hudi/pull/991#issuecomment-557029657), which was 15 days ago that I know of. I apologize if there is a notification I missed.. but I stepped in since vinoyang was blocked for more than some time.. We should formalize what the timeout here is as a group. 

> I'm not sure what is your confusion on the ownership here, which you have repeatedly mentioned.
Overall, I still don't see a JIRA with all sub-tasks. None of the PRs are tagged with a JIRA.. Is the one JIRA that @yanghua owns the one that this branch/PR is fixing?  I would like to understand all different follow ups here (may be your bullet list above is that) translated to jiras with assignees.. Has that happened? Apologize if I missed these?

yanghua commented:
I suggest we stop paying attention and explain the past, it is normal for distributed development to have such problems. We should focus on the future. We should maintain close and timely communication, synchronize our information and progress in a timely manner, and ensure that we are on the track of correct and rapid iteration.

I've started to verify the existing implementation based on @n3nash 's branch, let's go back to the previous branch(The price is not high, I did not go too far on another branch). 

@vinothchandar  In addition, there is currently no additional JIRA task to describe it except HUDI-289. I suggest that we still strictly follow the regular Jira and PR corresponding process to promote this feature. We can continue this rule even on feature branches.

We can create groups of this feature on Slack to facilitate our communication, even though we may not be able to use it under normal circumstances.

vinothchandar commented:
Yes. lets please streamline the tasks here.. We could also get help from others, if we clearly layout the work. Happy to help as needed.

yanghua commented:
I have made [HUDI-289](https://issues.apache.org/jira/browse/HUDI-289) as an umbrella issue and created some subtasks to measure detailed subtasks.

vinothchandar commented:
@yanghua thanks! So I take it that HUDI-394 is what @n3nash owns via fixes you mentioned?

n3nash commented:
@yanghua Thanks for the details, I've squashed the commits.

yanghua commented:
> @yanghua thanks! So I take it that HUDI-394 is what @n3nash owns via fixes you mentioned?

Yes, it's the first step of HUDI-289 and it blocks other subtasks. It has been done and just mapping the exists implementation provided by @n3nash .

yanghua commented:
> @yanghua Thanks for the details, I've squashed the commits.

Thanks, if you do not mind, I'd like to rename your squashed commit message and carry `HUDI-394` at the beginning of the commit message so that all the works can be tracked via Jira issues. It can provide a more detailed context for the community. WDYT?

n3nash commented:
@yanghua sounds good.
